src/libraries/PCPricing.sol

```diff
@@ -0,0 +1,303 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.19;
+
+/// @title Preferred Currency Pricing (PCPricing) Library
+/// @author Permissionless Games & ChatGpt
+/// @notice This library enables dynamic price adjustments based on preferred currency usage.
+/// @dev When the anchor currency is used, the price of all other tracked currencies can be decreased.
+///      A universal adjustment factor applies to all non-anchor currencies, promoting dynamic pricing models.
+
+library PCPricing {
+    using PCPricing for PCPricing.PricingData;
+    /// @notice Emitted when a new anchor currency is set
+    /// @param currency The currency that was set as the anchor
+    /// @param price The price of the anchor currency
+
+    event AnchorCurrencySet(bytes indexed currency, uint256 price);
+
+    /// @notice Emitted when the adjustment factor is updated
+    /// @param numerator The numerator of the adjustment factor
+    /// @param denominator The denominator of the adjustment factor
+    event AdjustmentFactorSet(uint256 numerator, uint256 denominator);
+
+    /// @notice Emitted when a new currency price is set
+    /// @param currency The currency that was set
+    /// @param price The price of the currency
+    event CurrencyPriceSet(bytes indexed currency, uint256 price);
+
+    /// @notice Emitted when a currency price is adjusted
+    /// @param currency The currency that was adjusted
+    /// @param newPrice The new price of the currency
+    /// @param increased Whether the price was increased or decreased
+    event CurrencyPriceAdjusted(
+        bytes indexed currency,
+        uint256 newPrice,
+        bool increased
+    );
+
+    /// @notice Emitted when a currency is removed
+    /// @param currency The currency that was removed
+    event CurrencyRemoved(bytes indexed currency);
+
+    /// @notice Emitted when all non-anchor prices are reduced
+    event NonAnchorPricesReduced();
+
+    /// @notice Emitted when a batch of non-anchor prices are adjusted
+    event NonAnchorPricesAdjustedBatch(
+        uint256 processedCount,
+        uint256 nextIndex
+    );
+
+    /// @notice Struct for pricing data
+    struct PricingData {
+        bytes anchorCurrency; // The anchor (base) currency
+        uint256 adjustmentNumerator; // The numerator of the universal adjustment percentage
+        uint256 adjustmentDenominator; // The denominator of the universal adjustment percentage
+        mapping(bytes => uint256) currencyPrice; // Mapping of currency prices
+        mapping(bytes => uint256) currencyIndex; // Mapping of currency index
+        bytes[] trackedCurrencies; // List of tracked non-anchor currencies
```

Comment:
Why `bytes[] trackedCurrencies`? What is the need for this? I always find it preferable to instead have `uint256 numPairedCurrencies` and `mapping(uint256 => bytes) pairedCurrency`. (Similar to OpenZeppelin `ERC721Enumerable`).